File: Railcar/RCSetupWindowController.rb
class RCSetupWindowController < NSWindowController
  attr_accessor :label, :errorLabel, :controlButton, :progressBar, :installer, :sourceCheckbox, :brewTitle, :brewInfo, :rbenvTitle, :rbenvInfo, :rubyTitle, :rubyInfo, :packagesTitle, :packagesInfo

  def windowDidLoad
    super
    
    if !installer.needsInstall?
      noInstallNeeded
    end
  end

  def windowWillClose(sender)
    NSApp.stopModal
  end

  def errorOccurred(message)
    errorLabel.setStringValue message
    progressBar.setDoubleValue 0.0
  end

  def compilerExists
    highlight(brewTitle, brewInfo)
    brewTitle.setStringValue "Installing Homebrew"

    progressBar.incrementBy 10.0
  end

  def brewInstalled
    fade(brewTitle, brewInfo)
    brewTitle.setStringValue "Homebrew installed"

    highlight(rbenvTitle, rbenvInfo)
    rbenvTitle.setStringValue "Installing RbEnv"

    progressBar.incrementBy 25.0
  end

  def rbenvInstalled
    fade(rbenvTitle, rbenvInfo)
    rbenvTitle.setStringValue "RbEnv installed"

    highlight(rubyTitle, rubyInfo)
    rubyTitle.setStringValue "Installing Ruby 1.9.3"

    progressBar.incrementBy 15.0
  end

  def rubyInstalled
    fade(rubyTitle, rubyInfo)
    rubyTitle.setStringValue "Ruby installed"

    highlight(packagesTitle, packagesInfo)
    packagesTitle.setStringValue "Installing default packages"

    progressBar.incrementBy 39.0
  end

  def gemsInstalled
    fade(packagesTitle, packagesInfo)
    packagesTitle.setStringValue "Default packages installed"

    progressBar.setDoubleValue 100.0
    progressBar.stopAnimation self

    controlButton.setTitle("Let's go!")
    controlButton.setEnabled(true)
    window.makeFirstResponder(controlButton)
  end

  def controlButtonClick(sender)
    if installer.needsInstall?
      installEverything
    else
      hideMe
    end
  end

  def noInstallNeeded 
    label.setStringValue "You're already setup.  Proceed!"
    progressBar.setDoubleValue 100.0
    progressBar.stopAnimation self
    
    controlButton.setEnabled false
  end

  def hideMe
    NSApp.stopModal
  end

  def installEverything
    fade(brewTitle, brewInfo, rbenvTitle, rbenvInfo, rubyTitle, rubyInfo, packagesTitle, packagesInfo)

    controlButton.setEnabled false
    sourceCheckbox.setEnabled false
    progressBar.performSelectorOnMainThread("startAnimation:", withObject:self, waitUntilDone:false)

    installer.performSelectorInBackground("installDependencies", withObject:(sourceCheckbox.state == 1))
  end

  def fade(*controls)
    controls.each {|control| control.setTextColor(NSColor.grayColor)}
  end

  def highlight(*controls)
    controls.each {|control| control.setTextColor(NSColor.blackColor)}
  end
end

File: Railcar/RCSetupWindowController.m
//
//  RCSetupWindowController.m
//  Railcar
//
//  Created by Jeremy McAnally on 3/29/12.
//  Copyright (c) 2012 Arcturo. All rights reserved.
//

#import "RCSetupWindowController.h"

@implementation RCSetupWindowController

@synthesize installer;
@synthesize setupWindow;

- (id)initWithWindow:(NSWindow *)window
{
    self = [super initWithWindow:window];
    if (self) {
        // Initialization code here.
    }
    
    return self;
}

- (void)windowDidLoad
{
    [super windowDidLoad];
    
    if ([installer needsInstall] == true) {
        [consoleButton setEnabled:NO];
    
        [setupWindow makeFirstResponder:installButton];
    } else {
        [self noInstallNeeded];
    }
}

-(void) errorOccurred:(NSString *)errorMessage {
    [label setStringValue:@""];
    [errorLabel setStringValue:errorMessage];
    [progressBar setDoubleValue:0.0];
}

-(void) compilerExists {
    [label setStringValue:@"Compiler found!  Installing brew..."];
    [progressBar incrementBy:10.0];
}

-(void) brewInstalled {
    [label setStringValue:@"Brew installed!  Installing rbenv..."];
    [progressBar incrementBy:25.0];
}

-(void) rbEnvInstalled {
    [label setStringValue:@"RbEnv installed!  Installing ruby... (This will take a few minutes!)"];
    [progressBar incrementBy:15.0];
}

-(void) rubyInstalled {
    [label setStringValue:@"Ruby installed!  Installing default gems..."];
    [progressBar incrementBy:39.0];
}

-(void) gemsInstalled {
    [label setStringValue:@"All setup!!"];
    [progressBar setDoubleValue:100.0];
    [progressBar stopAnimation:self];

    [[NSSound soundNamed:@"burn complete"] play];
    [consoleButton setEnabled:YES];
}

-(void) noInstallNeeded {
    [label setStringValue:@"You're already setup.  Proceed!"];
    [progressBar setDoubleValue:100.0];
    [progressBar stopAnimation:self];
    
    [consoleButton setEnabled:YES];
    [installButton setEnabled:NO];
}

-(IBAction)installEverything:(id)sender {
    [installButton setEnabled:NO];
    [progressBar performSelectorOnMainThread:@selector(startAnimation) withObject:self waitUntilDone:NO];

    [installer performSelectorInBackground:@selector(installDependencies) withObject:nil];
}

-(IBAction)openConsole:(id)sender {
    NSString * pathToInitializer = [NSString stringWithFormat:@"%@/homebrew/rbenv_init.sh", [[NSBundle mainBundle] bundlePath]];
    
    [[NSWorkspace sharedWorkspace] openFile:@"~/" withApplication:@"Terminal"];
    NSString *s = [NSString stringWithFormat:
                   @"tell application \"Terminal\" to do script \"source %@\"", pathToInitializer];
    
    NSAppleScript *as = [[NSAppleScript alloc] initWithSource: s];
    [as executeAndReturnError:nil];
}

@end
